File: day53/maxDepth.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if not root: return 0
        self.max_value = 0
        self._dfs(root, 0)
        return self.max_value
        
    def _dfs(self, node, level):
        if not node: 
            self.max_value = max(self.max_value, level)
            return
        
        self._dfs(node.left, level + 1)
        self._dfs(node.right, level + 1)

# BFS
# class Solution:
#     def maxDepth(self, root: TreeNode) -> int:
#         if not root: return 0

#         queue = collections.deque()
#         queue.append(root)
#         res = 0
#         while queue:
#             size = len(queue)
#             res += 1
#             for _ in range(size):
#                 cur = queue.popleft()
#                 if cur.left: queue.append(cur.left)
#                 if cur.right: queue.append(cur.right)
#         return res

# recursion
# class Solution:
#     def maxDepth(self, root: TreeNode) -> int:
#         if not root: return 0
#         return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))

File: day53/maxDepth.java
/*
104. 二叉树的最大深度
给定一个二叉树，找出其最大深度。
二叉树的深度为根节点到最远叶子节点的最长路径上的节点数。
说明: 叶子节点是指没有子节点的节点。

https://leetcode-cn.com/problems/maximum-depth-of-binary-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// BFS
class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) return 0;

        Queue<TreeNode> queue = new LinkedList<>();
        int level = 0;
        queue.offer(root);

        while (!queue.isEmpty()) {
            level++;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode cur = queue.poll();
                if (cur.left != null) queue.offer(cur.left);
                if (cur.right != null) queue.offer(cur.right);
            }
            
        }
        return level;
    }
}

// recursion
// class Solution {
//     public int maxDepth(TreeNode root) {
//         return root == null ? 0 : 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
//     }
// }
